Detectors/FOCAL/base/src/Geometry.cxx

```diff
@@ -0,0 +1,1633 @@
+                   << minlayer << ", " << maxLayer << ", " << padSize << ", " << isPixel << ")";
+      } // end if SEGMENT_LAYOUT
+
+    } // end if VIRTUAL
+
+  } // end while
+
+  setUpTowerWaferSize();
+  /////// re-arrange the longitudinal components
+  mNPixelLayers = npixlayers;
+  for (int i = 0; i < npixlayers; i++) {
+    mPixelLayerLocations[i] = pixl[i];
+  }
+
+  mNPadLayers = npadlayers;
+  LOG(debug) << "mNPadLayers, mNPixelLayers, mNHCalLayers, mNumberOfSegments :: " << mNPadLayers << " / " << mNPixelLayers << " / " << mNHCalLayers << " / " << mNumberOfSegments;
+
+  mLayerSeg = (mNPadLayers + mNPixelLayers + mNHCalLayers) / mNumberOfSegments;
+
+  if (mNumberOfSegments >= 100) {
+    LOG(warning) << "You reached the segments limits! Setting Number of segments to: 100";
+    mNumberOfSegments = 99;
+    LOG(warning) << "New number of segments: " << mNumberOfSegments;
+    mLayerSeg = (mNPadLayers + mNPixelLayers + mNHCalLayers) / mNumberOfSegments;
+  }
+  if ((mNPadLayers + mNPixelLayers + mNHCalLayers) % mNumberOfSegments) {
+    mNumberOfSegments++;
+    for (int i = 0; i < mNumberOfSegments; i++) {
+      mNumberOfLayersInSegments[i] = mLayerSeg;
+    }
+    LOG(debug) << "Number of segments: " << mNumberOfSegments;
+  } else {
+    for (int i = 0; i < mNumberOfSegments; i++) {
+      mNumberOfLayersInSegments[i] = mLayerSeg;
+    }
+  }
+
+  setUpLayerSegmentMap();
+
+  float center_z = 0;
+
+  mPadCompositionBase.reserve(200);
+  mHCalCompositionBase.reserve(200);
+  mPixelCompositionBase.reserve(200);
+  mFrontMatterCompositionBase.reserve(200);
+
+  for (auto& tmpComp : padCompDummy) {
+    LOG(debug) << "Material(pad layer) " << tmpComp.Material();
+    LOG(debug) << "layer / stack / id :: " << tmpComp.Layer() << " / " << tmpComp.Stack() << " / " << tmpComp.Id();
+    LOG(debug) << "center x,y,dz :: " << tmpComp.SizeX() << " / " << tmpComp.SizeY() << " / " << tmpComp.SizeZ();
+    if (tmpComp.Material().compare("SiPad")) { // materials other than SiPad
+      mPadCompositionBase.push_back(*(new Composition(tmpComp.Material(),
+                                                      tmpComp.Layer(), tmpComp.Stack(), tmpComp.Id(),
+                                                      tmpComp.CenterX(), tmpComp.CenterY(), center_z,
+                                                      mTowerSizeX, mTowerSizeY, tmpComp.SizeZ())));
+      if (mTowerSizeX < tmpComp.SizeX()) {
+        mTowerSizeX = tmpComp.SizeX();
+      }
+      if (mTowerSizeY < tmpComp.SizeY()) {
+        mTowerSizeY = tmpComp.SizeY();
+      }
+    } else {
+      for (int itowerX = 0; itowerX < mGlobal_PAD_NX_Tower; itowerX++) {
+        for (int itowerY = 0; itowerY < mGlobal_PAD_NY_Tower; itowerY++) {
+          for (int ix = 0; ix < mGlobal_PAD_NX; ix++) {
+            for (int iy = 0; iy < mGlobal_PAD_NY; iy++) {
+              auto padCenter = getGeoPadCenterLocal(itowerX, itowerY, iy, ix);
```

Comment:
You could use structured binding to access the content of the tuple directly. Makes the code a bit better readable.